snapcraft/parts/plugins/kernel_build.sh

```diff
@@ -0,0 +1,323 @@
+check_config() {
+  echo "$_required_boot"     \
+       "$_required_generic"  \
+       "$_required_security" \
+       "$_required_snappy"   \
+       "$_required_systemd"  | while read -r config; do
+    if ! grep "^CONFIG_${config}=" "${CRAFT_PART_BUILD}/.config"; then
+      printf '*** WARNING ***\n'
+      printf 'Your kernel config is missing:\n'
+      printf '%s\n' "$config"
+      printf 'Which is recommended or required by Ubuntu Core!\n'
+      printf 'Continuing, but you should check your .config\n'
+    fi
+  done
+}
+
+# Gather kernel release info to encode in kernel artifacts
+release_info() {
+  printf 'Gathering release information\n'
+  DEBIAN="${CRAFT_PART_SRC}/debian"
+  src_pkg_name=$(sed -n '1s/^\(.*\) (.*).*$/\1/p'                                    "${DEBIAN}/changelog")
+  release=$(     sed -n '1s/^'"${src_pkg_name}"'.*(\(.*\)-.*).*$/\1/p'               "${DEBIAN}/changelog")
+  revision=$(    sed -n '1s/^'"${src_pkg_name}"'\ .*('"${release}"'-\(.*\)).*$/\1/p' "${DEBIAN}/changelog")
+  abinum=${revision%.*}
+  uploadnum=${revision##*.}
+  abi_release="${release}-${abinum}"
+}
+
+fetch_zfs() {
+  printf 'Cloning ZFS for %s\n' "$UBUNTU_SERIES"
+
+  if [ ! -d "${CRAFT_PART_BUILD}/zfs" ]; then
+    git clone \
+      --depth 1 \
+      --branch "applied/ubuntu/$UBUNTU_SERIES" \
+      https://git.launchpad.net/ubuntu/+source/zfs-linux \
+      "${CRAFT_PART_BUILD}/zfs"
+  fi
+}
+
+build_zfs() {
+  printf 'Building zfs modules...\n'
+  cd "${CRAFT_PART_BUILD}/zfs"
+
+  ./configure \
+    --with-linux="$CRAFT_PART_SRC"         \
+    --with-linux-obj="$CRAFT_PART_BUILD"   \
+    --host="$CRAFT_ARCH_TRIPLET_BUILD_FOR" \
+    --with-config=kernel
+
+  make -j "$CRAFT_PARALLEL_BUILD_COUNT"
+  make install DESTDIR="${CRAFT_PART_INSTALL}/zfs"
+
+  mv -f "${CRAFT_PART_INSTALL}/zfs/lib/modules/${kver}/extra" \
+        "${CRAFT_PART_INSTALL}/modules/${kver}"
+
+  rm -rf "${CRAFT_PART_INSTALL}/zfs"
+}
+
+build_perf() {
+  printf 'Building perf binary...\n'
+
+  mkdir -p "${CRAFT_PART_BUILD}/tools/perf"
+
+  # Override source and build directories
+  make -j "$CRAFT_PARALLEL_BUILD_COUNT"  \
+       -C "${CRAFT_PART_SRC}/tools/perf" \
+        O="${CRAFT_PART_BUILD}/tools/perf"
+
+  install -Dm0755 "${CRAFT_PART_BUILD}/tools/perf/perf" "${CRAFT_PART_INSTALL}/bin/perf"
+}
+
+redepmod() {
+  printf 'Rebuilding module dependencies\n'
+  depmod -b "$CRAFT_PART_INSTALL" "$kver"
+}
+
+# Cleanup previous builds
+if [ -e "${CRAFT_PART_INSTALL}/modules" ] ||\
+   [ -L "${CRAFT_PART_INSTALL}/lib/modules" ]; then
+  cleanup
+fi
+
+### Setup
+# Create new config if one does not exist
+[ -e "${CRAFT_PART_BUILD}/.config" ] || {
+  # Privilege a specified flavour over all else
+  if [ -n "$flavour" ] && [ "$flavour" != "generic" ]; then
+      printf "Using Ubuntu config flavour %s\n" "$flavour"
+      gen_flavour_config
+    # Privilege specified config(s) over most
+    elif [ -n "$defconfig" ] && [ "$defconfig" != "defconfig" ]; then
```

Comment:
`$defconfig` is `list[str]` so we'd need to check *every* element, or at least that `$defconfig` is multiple elements (if you list `defconfig kvmconfig`, we should privilege that over `flavour=generic` and run with it).